Update targets delegations in generate_targets_metadata

sechkova opened the pull request:
**Fixes issue #**: #1037

**Description of the changes being introduced by the pull request**:
- Collect keys and threshold of delegated roles and update delegations in `generate_targets_metadata() `in a similar manner as `generate_root_metadata()` does for top-level roles.
- Move key dictionary generation from `generate_root_metadata()` to a helper function` _add_keys_to_keydict()` 
- Update tests

Fixing this issue was partially blocked by #574: after loading the repository the delegated roles' 'keyids' and 'threshold' are lost which leads to throwing "missing key" exceptions while trying to update the delegations.

I suggest a short-term solution of #574 by using the recently added logic of hierarchically loading the delegations in `load_reposiroty()`. This makes it possible to get the delegated roles `keyids` and `threshold` from its delegating role metadata.
Clearly this won't work if more than one delegation to the same role exists but ... this is not supported correctly anyway in the current implementation so lets cover the simplest use case here?

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
Move _keys_to_keydict() to repository_lib

Use _keys_to_keydict() for the  key dictionary generation in
generate_root_metadata().
Rename it as a public function keys_to_keydict().

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Update delegations in generate_targets_metadata()

Collect keys and threshold of delegated roles and update
delegations in generate_targets_metadata in a similar manner
as generate_root_metadata() does for top-level roles.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Load delegated roles 'keyids' and 'threshold'

Use the delegation graph traversal during load_repository()
to load delegated roles' 'keyids' and 'threshold' by reading it
from the delegating role metadata.

If more than one delegation to the same role exists, only the first
one is loaded in roledb for this role.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Update failing tests for generate_targets_metadata

Tests logic is modified to accommodate for the update of the
delegations during generate_targets_metadata().

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add test for delegations update

Add a new test case in test_generate_targets_metadata
to check if targets metadata is up-to-date with its delegated
roles.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -155,9 +155,16 @@ def _generate_and_write_metadata(rolename, metadata_filename,
+        storage_backend, repository_name)
+
+    # Update roledb with the latest delegations info
+    if delegations_update:
```
I assume the reason we're having `generate_targets_metadata()` return a boolean value and testing it here is because `update_roleinfo()` is non-performant due to its use of `deepcopy()`? What is the impact if we just `update_roleinfo()` without this check?

sechkova replied:
Yes, the idea is to avoid calling `update_roleinfo()` for each delegation on every `write`, especially when their number can grow high if delegated hashed bins are used. 

Anyway: 
- when tested with the scrip from https://github.com/theupdateframework/tuf/issues/1005#issuecomment-601351054 the overhead when updating all delegations and not using the `delegations_update` boolean seems to be not more than ~1-2secs (tested in a regular VM on my laptop)
- assuming that delegated hashed bins all use the same key, key revocation would mean updating all delegations

So if you are hinting towards simplifying the code at the cost of some negligible efficiency, you may have a point :slightly_smiling_face: 

In general, the overall overhead of the code in this PR seems to be in the range of 1-2 seconds (too small to actually calculate an exact delta without a dedicated test setup)

joshuagl replied:
Thanks for crunching the numbers. I'd absolutely be in favour of the code simplification.

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1466,6 +1468,47 @@ def generate_targets_metadata(targets_directory, target_files, version,
+
+  for keyid in keyids:
+
+    # Do we have a duplicate?
+    if keyids.count(keyid) > 1:
+      raise securesystemslib.exceptions.Error('Same keyid listed twice:'
+          ' ' + keyid)
```
This feels odd. We do not handle this exception anywhere, is this likely to happen in practice?
Furthermore, this is all data we manage. If duplicates really are an issue, should we just prune them? What are we expecting the 'user' to do when they see this exception?

sechkova replied:
There is a test case about it :grin: 
In practice `add_verification_key` has a check to avoid duplicates.
For the rest of your questions, I don't have a strong opinion but I agree this does not seem to be the right place of the check. Unfortunately,  I cannot claim that I can track all possible data entry points for roledb  ... while this is the only "exit" point.  Maybe this is the historical reason for it. 
I would gladly remove the check and its test but I don't have a great insight on what to put instead. Nothing?

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1401,9 +1381,31 @@ def generate_targets_metadata(targets_directory, target_files, version,
+    for role in delegations['roles']:
+      rolename = role['name']
+      role_keyids = tuf.roledb.get_role_keyids(rolename, repository_name)
+      role_threshold= tuf.roledb.get_role_threshold(rolename, repository_name)
```
```suggestion
      role_threshold = tuf.roledb.get_role_threshold(rolename, repository_name)
```

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1401,9 +1381,31 @@ def generate_targets_metadata(targets_directory, target_files, version,
+        role['threshold'] = role_threshold
+        delegation_update = True
+
+      _add_keys_to_keydict(keydict, role_keyids, repository_name)
```
Should this be part of the `if` conditional?

sechkova replied:
No, it has to be outside the `if`. The key dictionary has to contain delegations keys of all roles, updated or not.

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1401,9 +1381,31 @@ def generate_targets_metadata(targets_directory, target_files, version,
+
+      _add_keys_to_keydict(keydict, role_keyids, repository_name)
+
+    if delegation_update:
+      delegations['keys'] = keydict
```
Is there any harm in always assigning keydict to `delegations['keys']`? So far as I can tell `delegations['keys']` is only used when `delegation_update == True`?

sechkova replied:
No, no obvious harm ...

joshuagl reviewed (commented):
Thanks for the PR @sechkova! Your approach seems sound, but I have a couple of questions in-line about the implementation.

jku commented on tuf/repository_lib.py:
```diff
@@ -1466,6 +1431,47 @@ def generate_targets_metadata(targets_directory, target_files, version,
+    if keyid not in keydict:
+
+      # This appears to be a new keyid.  Generate the key for it.
+      if key['keytype'] in ['rsa', 'ed25519', 'ecdsa-sha2-nistp256']:
```
I wonder if there's a reason for hard coding the key types here: we could just let format_keyval_to_metadata() throw FormatError (or handle it and throw something better) ?

sechkova replied:
Thanks for pointing it :) 
After a quick search,  `format_keyval_to_metadata()` is used at a couple of places in `tuf` the way you described it (just letting it throw FormatError) so I agree applying the same here.

It seems also worth trying to merge the functionality of `repository_tool._keys_to_keydict()` and `repository_lib._add_keys_to_keydict()` which seem suspiciously similar :eyes:

sechkova committed:
Fix typo in comment section

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Some time has passed since the last updates on this pull request so I had to rebase and force push. However I think the reformatted set of changes is addressing all the comments up to now.

joshuagl commented:
> Some time has passed since the last updates on this pull request so I had to rebase and force push. However I think the reformatted set of changes is addressing all the comments up to now.

Thank you for the updates. They look good on a first pass, I'm aiming to do a more thorough review next week.

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -2256,6 +2228,27 @@ def disable_console_log_messages():
+  """
+  Iterate over a list of keys and return a list of keyids and a dict mapping
+  keyid to key metadata
+  """
```
As we have promoted this to a public method (implicitly, by removing the leading `_`) we should improve the docstring so that it's easier for callers to make use of. Docstrings for public methods should include `<Arguments>` and `<Returns>` sections, at least. The `Arguments` section for this method should help a reader understand the types and any constraints of the objects in the list.

sechkova replied:
Done in  1acd07c

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1365,6 +1376,9 @@ def generate_targets_metadata(targets_directory, target_files, version,
     'tuf.formats.TARGETS_SCHEMA'.
+
+    delegation_update:
+      Boolean set to True if delegations are updated.
```
I think this value is no longer returned?

sechkova replied:
Oops ... 
Removed in a18beb2

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1382,8 +1396,25 @@ def generate_targets_metadata(targets_directory, target_files, version,
         ' targets and using existing fileinfo.')
 
-  if delegations is not None:
+  if delegations:
```
Can you help me understand this change? 

`if delegations is not None` ensures we only enter this loop if the delegations argument is changed from its default value of `None`.  As I understand it this is the standard pattern for checking whether an argument with a default value of `None` has been changed.

sechkova replied:
My bad ... I didn't think it through ...  :facepalm:  
Reverted in a18beb2

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -961,6 +942,21 @@ def test__load_top_level_metadata(self):
+    delegations = targets_signable['signed']['delegations']
+
+    roleinfo = {}
+    roleinfo['name'] = 'role1'
```
Nit: it's probably better read the role name from the loaded data in `delegations`, as we are for keyids and threshold, in case the generated data changes.

sechkova replied:
Thanks, done in f1a0fb1

joshuagl reviewed (changes requested):
Thank you. I have a few minor comments, mostly on docstring, but otherwise this is looking good.

sechkova commented:
Someone with enough permissions to restart travis-ci? Seems like another timeout 👀

jku commented:
Bad news is  #1096 was merged and this now has conflicts. Good news is the proxy tests _should_ be more reliable once you rebase :)

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1412,6 +1395,23 @@ def generate_targets_metadata(targets_directory, target_files, version,
+    delegations_keys = []
+    # Update 'keyids' and 'threshold' for each delegated role
+    for role in delegations['roles']:
+      role['keyids'] = \
+          tuf.roledb.get_role_keyids(role['name'], repository_name)
+      role['threshold']= \
+          tuf.roledb.get_role_threshold(role['name'], repository_name)
```
The [coding style](https://github.com/secure-systems-lab/code-style-guidelines#code-lay-out) suggests:
> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. If necessary, you can add an extra pair of parentheses around an expression, but sometimes using a backslash looks better.

Which suits my personal preference too.
```suggestion
      role['keyids'] = tuf.roledb.get_role_keyids(role['name'],
          repository_name)
      role['threshold']= tuf.roledb.get_role_threshold(role['name'],
          repository_name)
```

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -2256,9 +2256,42 @@ def disable_console_log_messages():
 if __name__ == '__main__':
   # The interactive sessions of the documentation strings can
-  # be tested by running repository_tool.py as a standalone module:
+  # be tested by running repository_lib.py as a standalone module:
```
👍

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -387,6 +369,34 @@ def test_generate_targets_metadata(self):
+    delegations = roleinfo['delegations']
+    old_delegations = copy.deepcopy(delegations)
+
+    targets_metadata  = \
+      repo_lib.generate_targets_metadata(targets_directory, target_files,
+                                         version, expiration_date, delegations,
+                                         False)
```
I realise you've just copied other examples in this file, but it's more inline with the coding style to do something like:
```suggestion
    targets_metadata  = repo_lib.generate_targets_metadata(targets_directory,
        target_files, version, expiration_date, delegations, False)
```

joshuagl reviewed (approved):
Thanks for the updates @sechkova, this looks much nicer.

I have made a few minor style suggestions and this PR needs rebasing on the most recent changes in develop, but otherwise this looks good to me.

sechkova commented:
Thanks @joshuagl, applied your code style suggestions and rebased. Let's see the tests :) 

_Edit:_ Two of the tests fail due to reasons unknown to me ... would need another look ... seems related to https://github.com/secure-systems-lab/securesystemslib/pull/266

jku commented:
lint failure is #1115

sechkova commented:
Rebased on develop with the linter failure fix.

joshuagl commented on tuf/repository_lib.py:
```diff
@@ -1238,43 +1239,11 @@ def generate_root_metadata(version, expiration_date, consistent_snapshot,
 
-    # Keep track of the keys loaded to avoid duplicates.
-    keyids = []
-
-    # Generate keys for the keyids listed by the role being processed.
-    for keyid in tuf.roledb.get_role_keyids(rolename, repository_name):
+    # Collect keys from all roles in a list
+    keyids = tuf.roledb.get_role_keyids(rolename, repository_name)
+    for keyid in keyids:
```
Nit: this change to assigning the return value of `roledb.get_role_keyids` seems superfluous, is `keyids` used anywhere other than the iterator here? Why not just keep for `keyid in tuf.roledb....`?

sechkova replied:
Yes, it is used in https://github.com/theupdateframework/tuf/pull/1074/files#diff-9f585d5478a17bd9e0f6cd5f7484ab60L1284

It replaces the keyids list that was used to avoid duplicates which we already decided not necessary somewhere in the discussion above.

joshuagl replied:
Oh, yes I see. Thanks!

joshuagl commented:
Good work, thanks!
